base/iobuffer.jl

```diff
@@ -489,6 +489,86 @@ function take!(io::IOBuffer)
     return data
 end
 
+"Internal method. This method can be faster than takestring!, because it assumes
+the buffer is writable and seekable, and because it does not reset the buffer
+to a usable state.
+Using the buffer after calling unsafe_takestring! may cause illegal behaviour.
+This function is meant to be used when the buffer is only used as a temporary
+string builder, which is discarded after the string is built."
+function unsafe_takestring!(io::IOBuffer)
+    off = io.offset
+    nbytes = io.size - off
+    iszero(nbytes) && return ""
+    # The C function can only copy from the start of the memory.
+    # Fortunately, in most cases, the offset will be zero.
+    return if iszero(off)
+        ccall(:jl_genericmemory_to_string, Ref{String}, (Any, Int), io.data, nbytes)
+    else
+        mem = StringMemory(nbytes)
+        unsafe_copyto!(mem, 1, io.data, off + 1, nbytes)
+        unsafe_takestring!(mem)
+    end
+end
+
+"""
+    takestring!(io::IOBuffer) -> String
+
+Return the content of `io` as a `String`, resetting the buffer to its initial
+state.
+This is preferred over calling `String(take!(io))` to create a string from
+an `IOBuffer`.
+
+# Examples
+```jldoctest
+julia> io = IOBuffer();
+
+julia> write(io, [0x61, 0x62, 0x63]);
+
+julia> s = takestring!(io)
+"abc"
+
+julia> isempty(take!(io)) # io is now empty
+true
+```
+
+!!! compat "Julia 1.12"
+    This function requires at least Julia 1.12.
+"""
+function takestring!(io::IOBuffer)
+    nbytes = filesize(io)
+
+    # If the memory is invalidated (and hence unused), or no bytes, return empty string
+    return if (iszero(nbytes) || io.reinit)
```

Comment:
Can't this code call `unsafe_takestring!(io)`?

Reply:
Re: This comment and the one below: I couldn't remember anymore, and had to re-read the `iobuffer.jl` code. I must admit I find it difficult to understand what invariants the fields of the buffer follow - the precise meaning of the fields is obscure.
Nonetheless, it looks to me like the difference between this computation in `takestring!` and `unsafe_takestring!` is because the latter assumed the input buffer is seekable, which apparently changes whether the beginning of the readable data is given by `.ptr` or `.offset + 1`.
However, it looks like `IOBuffer` is always seekable, since all the public constructors only return instances with `seekable=true`.

So, in the latest commit, I indeed just call `unsafe_takestring!` here, using the same computation.
Let's see if the tests pass.